src/main/scala/mesosphere/marathon/tasks/PortsMatcher.scala

```diff
@@ -47,145 +50,213 @@ class PortsMatcher(app: AppDefinition, offer: Offer, acceptedResourceRoles: Set[
+      val builder = Seq.newBuilder[protos.Range]
+
+      @tailrec
+      def process(lastRangeOpt: Option[protos.Range], next: Seq[PortWithRole]): Unit = {
+        (lastRangeOpt, next.headOption) match {
+          case (None, _) =>
+          case (Some(lastRange), None) =>
+            builder += lastRange
+          case (Some(lastRange), Some(nextPort)) if lastRange.end == nextPort.port - 1 =>
+            process(Some(lastRange.copy(end = nextPort.port.toLong)), next.tail)
+          case (Some(lastRange), Some(nextPort)) =>
+            builder += lastRange
+            process(Some(nextPort.toRange), next.tail)
+        }
+      }
+      process(ranges.headOption.map(_.toRange), ranges.tail)
 
-          RangesResource(Resource.PORTS, portRanges, role)
-      }.to[Seq]
+      builder.result()
     }
 
-}
+    /**
+      * We want to avoid reusing the same dynamic ports. Thus we shuffle the ports to make it less likely.
+      *
+      * The implementation idea:
+      *
+      * * Randomize the order of the offered ranges.
+      * * Now treat the ports contained in the ranges as one long sequence of ports.
+      * * We randomly choose an index where we want to start assigning dynamic ports in that sequence. When
+      *   we hit the last offered port with wrap around and start offering the ports at the beginning
+      *   of the sequence up to (excluding) the port index we started at.
+      */
+    def randomPortsFromRanges(random: Random = Random)(offeredPortRanges: Seq[PortRange]): Iterator[PortWithRole] = {
+      val numberOfOfferedPorts = offeredPortRanges.map(_.scalaRange.size).sum
+
+      if (numberOfOfferedPorts == 0) {
+        return Iterator.empty
+      }
+
+      def findStartPort(shuffled: Vector[PortRange], startPortIdx: Int): (Int, Int) = {
+        var startPortIdxOfCurrentRange = 0
+        val rangeIdx = shuffled.indexWhere {
+          case range: PortRange if startPortIdxOfCurrentRange + range.scalaRange.size > startPortIdx =>
+            true
+          case range: PortRange =>
+            startPortIdxOfCurrentRange += range.scalaRange.size
+            false
+        }
+
+        (rangeIdx, startPortIdx - startPortIdxOfCurrentRange)
+      }
+
+      val shuffled = Random.shuffle(offeredPortRanges).toVector
+      val startPortIdx = Random.nextInt(numberOfOfferedPorts)
+
+      val (rangeIdx, portInRangeIdx) = findStartPort(shuffled, startPortIdx)
+      val startRangeOrig = shuffled(rangeIdx)
+
+      val startRange = startRangeOrig.withoutNPorts(portInRangeIdx)
+      val afterStartRange = shuffled.slice(rangeIdx + 1, shuffled.length).flatMap(_.portsWithRolesIterator)
+      val beforeStartRange = shuffled.slice(0, rangeIdx).flatMap(_.portsWithRolesIterator)
+      val endRange = startRangeOrig.firstNPorts(portInRangeIdx)
+
+      startRange ++ afterStartRange ++ beforeStartRange ++ endRange
+    }
+  }
+
+  case class PortRange(role: String, range: Protos.Value.Range) {
```

Comment:
Would it be better to make `range` a scala range with an apply in the PortRange object, that accepts the protobuf?